### PR TITLE
Auto-format source code with clang-format (config file)

### PR DESCRIPTION
**Pull request by uklotzde**
I've created a minimal `.clang-format` file based on the _Google_ C++ style and our style guide. Only those options that differ from the default _Google_ C++ style are included to keep it short and concise.

This initial configuration file can be tuned and adapted while we gradually reformat more files. Options for other languages like Python and JavaScript can be added on demand. We don't need to reformat all files at once, since that would introduce many merge conflicts!

The `.clang-format` file is picked up by _VS Code_ when opening the project folder and I now get correctly formatted code when saving files after editing them :)

### CODE REVIEW

File: src/track/beatgrid.cpp
Comment (daschuer) [inline src/track/beatgrid.cpp]: We use no space after // for deactivate code.

**Review by daschuer — CHANGES_REQUESTED** [review]
This PR also contains non clang-firmat related changes.

On one hand it is a good idea to have a common formating setting, on the other hand clang-format is quite aggressive to reformat every single line.

Clang is designed for not allowing two opinions when formating the code.
IMHO this does not fit to Mixxx current approach which allows alternatives.

It is perfectly reasonable to use clang-format to format touched lines, but I am afraid adding this file to the project root will encourage other developers to reformat whole files, which introduce just unnecessary noise. 

In this PR it is the indent change of the switch statement and the re-layout of the comments block. 

We are already short in reviewing capacity in our team and I have no interests to review unnecessary and sometimes stupidest clang changes.

Currently I am satisfied with the current styling state of the code. So what could be the way to get the best of both solutions.

**Comment by uklotzde** [comment]
I've reverted all manual changes.

**Comment by uklotzde** [comment]
Open for discussion.

File: src/track/beatgrid.cpp
Comment (uklotzde) [reply src/track/beatgrid.cpp]: Clang-Format does not distinguish if end-of-line comments contain code or arbitrary text.

**Comment by ywwg** [comment]
On every project I've worked on where this transition occurred, the volume of worried discussion pre-auto-formatting dwarfs the isolated problems that crop up post-auto-formatting.  There are initial cases where the autoformatter mangles some comments, but after that, it Just Works and people move on (with occasional grumbling about what the formatter chooses for certain long lines that should probably be rewritten anyway). Removing formatting from the list of discussion topics truly frees up an enormous amount of cognitive load.  Highly +1 using clang-format.

The question about creating noise per-PR is somewhat valid, and there are two good solutions:
1) when editing a file, if clang-format causes a lot of confusing diffs, create two commits -- one noop clang-format, and another with the actual changes
2) create a one-time single noop clang-format PR over the entire codebase.  This is what I would recommend.

**Comment by daschuer** [comment]
2 would work for me, if we move to plain Google style. In earlyer attempts it turns out that we get bad results, if we try to teach clang-format our current style.

But we have dozens of open PRs ...

**Comment by ywwg** [comment]
> But we have dozens of open PRs ...

I think open PRs can do:
1) run clang-format on their own PR
2) pull in formatted-HEAD

Then there shouldn't even be conflicts and the diffs will be clean

**Comment by daschuer** [comment]
This will result in chis cross merges where changes the same changes are done in the PR and in master and the origin is unclear. 

What is the actual issue we like to fix? 

I think it is primary the question how to setup VS Code with Mixxx. 
Does the VS code setting "editor.formatOnType": true work and only touch edited lines? 
If this works, this would be a good short hand here.

File: src/track/beatmap.cpp
Comment (daschuer) [inline src/track/beatmap.cpp]: This clang behavior is just bad. Sometimes we break lines earlier to structure the comment. Clang just reverts this.

File: src/track/beatutils.cpp
Comment (daschuer) [inline src/track/beatutils.cpp]: Does clang actually sort includes or is this a leftover manual change?

File: src/track/beatutils.cpp
Comment (daschuer) [inline src/track/beatutils.cpp]: Missing braces.

File: src/track/beatutils.cpp
Comment (daschuer) [inline src/track/beatutils.cpp]: Disabled code without a space.

**Comment by uklotzde** [comment]
This is 2k19, I don't want to care about code formatting any more. All code that I contribute in the future will be auto-formatted without exceptions. If anyone has a different or better auto-formatting solution, please provide it. Continuing without a tooling solution is not an option for me.

The version-controlled configuration file serves as a provable reference to avoid unnecessary and exhausting discussions. Tweaking the contents of this single file is much more effective than doing it again and again for individual source files.

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/track/beatutils.cpp
Comment (uklotzde) [reply src/track/beatutils.cpp]: clang-format is not able to check for this. For this purpose you need clang-tidy and set [readability-braces-around-statements](https://clang.llvm.org/extra/clang-tidy/checks/readability-braces-around-statements.html) to 0.

File: src/track/beatutils.cpp
Comment (uklotzde) [reply src/track/beatutils.cpp]: clang-format is not able to parse the text in comments and make a decision

Comment (uklotzde) [reply src/track/beatutils.cpp]: Commented out code should be deleted anyway.

File: src/track/beatutils.cpp
Comment (uklotzde) [reply src/track/beatutils.cpp]: Includes are sorted, default option of the google style.

File: src/track/beatmap.cpp
Comment (uklotzde) [reply src/track/beatmap.cpp]: I will set ReflowComments: false

File: src/track/beatmap.cpp
Comment (uklotzde) [reply src/track/beatmap.cpp]: I will set BreakConstructorInitializers: BeforeComma (now: BeforeColon)

File: src/track/beatutils.cpp
Comment (uklotzde) [reply src/track/beatutils.cpp]: I will set
* AllowShortBlocksOnASingleLine: true
* AllowShortCaseLabelsOnASingleLine: true
to avoid reformatting existing code

File: src/track/beatmap.cpp
Comment (uklotzde) [reply src/track/beatmap.cpp]: I will set IndentCaseLabels: false

Comment (daschuer) [reply src/track/beatmap.cpp]: We do not have our own rule about case labels. If the Google default is to indent them, it is OK. 
It is just that I assume we have both in the code so "don't care would be the best option" to avoid noise.

File: src/track/beatutils.cpp
Comment (daschuer) [reply src/track/beatutils.cpp]: This is IMHO an expedition, since it allows to reuse the debug statement during later development.

File: src/track/beatutils.cpp
Comment (uklotzde) [reply src/track/beatutils.cpp]: AllowShortBlocksOnASingleLine: true is not a good idea, because it will collapse even multi-line statements with braces into a single line.

File: src/track/beatmap.cpp
Comment (uklotzde) [reply src/track/beatmap.cpp]: We need to stick to BeforeColon

**Comment by uklotzde** [comment]
There will always be edge cases and some of our rules like the constructor initializer list indentation is simply not supported (because it uncommon, breaking before both colon and comma would be a consistent and supported alternative).

**Comment by rryan** [comment]
I don't have much to add beyond what @ywwg and @uklotzde have already said, but I'm also heavily in favor. There'll never be a good time to do this, and we'll always have tons of open PRs. I think we should tear the band-aid off. I personally won't mind fixing all the conflicts in my open PRs because I know it's for a greater good :).

**Comment by daschuer** [comment]
I think we all agree that an auto formater solution would be a benefit in our project.

The bullet point here is, what will be the effort and the smothest way to get this in.
Here some steps:

1.) Adjust our style guide: Since no ideal autoformater is available for out current style guide. We need to adjust it to be able to be done aromatically.
This PR already identifies some issues.
The result will be that the autoformater in a specific version becomes the final reference and not the guide. In case of clang-format this applies even if it suffer bugs since it touches every single line.

2.) Clean up PRs: It would be OK for me to autoformat touched lines only. This can be done with clang-format-diff or with IDE settings like ```"editor.formatOnType": true``` in VS Code.
A minor amount of unrelated changes in the neighborhood are still OK for me as usual. No change to the current workflow.

3.) Install a GitHub check that checks the PR diff only for clamg-format issues 

4.) Mass format all code:
We should not mass format all pending PRs because that produces criss-cross merges, which makes digging for changes with blame hard. The following will work IMHO best:
a.) Merge the target branch into all pending commits, and resolve conflicts if there are any. This ensures we have to deal with format issues only in g.)
b.) format 2.1 branch. 
c.) Merge it into 2.2
d.) Format 2.2 branch
e.) Merge it into master
f.) Format master
g.) Merge the target branches into the pending PRs and resolve conflicts.
I assume that this will be not too hard, if we pick a clang-format setting that close to our current style.

I suggest that all this is done by a single person and as pull request to the original PR branches.

I would be satisfied if we only do step 1. to 3.

Step 4. is a lot of effort and has a low priority for me compared to the GL issues and other pending PRs.
I have no objections if two people  can effort the time (developer and reviewer), but maybe we should review and merge finished PRs first. 

What do you think?

**Comment by ywwg** [comment]
thanks @uklotzde for twiddling the knobs to help keep our style as consistent as we can!  

Hopefully at the end of this the style guide around code formatting is just "use this settings file in clang-format".  @daschuer I think that's a good plan.  How much longer do we plan to support 2.1/2.2? I agree that the mass format will make those backports increasingly difficult, but I assume we don't plan on doing that forever.  (2.1 seems to be a special case because it's the last pre-qt5 release, am I right?)

**Comment by daschuer** [comment]
I think we can drop 2.1 support after the we have a solution for the Mac-Os GUI performance issue. Hopefully the pending 2.1.8 is the last one.

**Comment by uklotzde** [comment]
1) Agreed. First we should get our style guide and the tooling into a consistent state. Sticking to rules that are not supported by our recommended toolchain does not make any sense.

2) During the next phase developers may optionally auto-format their code manually with the recommended tools and settings. This is an opportunity to tune and adjust both the style guide and the toolchain configuration.

3 + 4) The installation of automatic checks and finally mass reformatting of the code base should be deferred until we are absolutely sure that everything works as expected and the auto-formatter produces the desired outcome.

5) In a last step we can remove all redundant recommendations from the style guide. The style guide does only need to contain those rules, that are not (yet) covered by auto-formatting to avoid inconsistencies.

**Comment by daschuer** [comment]
I agree if you add to 2. Do no mass-format PRs, auto format of touched lines are OK (clang-format-diff)

**Comment by daschuer** [comment]
We have discovered here two issues: 
* The fixed 100 columns rule versus 80 + X
I have tried to achieve this by using penalties, but this seams to have no effect. We may consider to set the column limit to 0 disable this feature. But line break is a core responsibility of an autoformater. 
* no space after // for commented code 
This id relevant for testing and debugging code. We may drop this rule or alternative surround by 
#ifdef TESTING similar, a pile of extra work.

**Comment by rryan** [comment]
> * no space after // for commented code
>   This id relevant for testing and debugging code. We may drop this rule or alternative surround by
>   #ifdef TESTING similar, a pile of extra work.

I think it's best to not submit commented code, and this style guide rule is intended to prevent people from doing it... but you can configure most editors to insert a space when commenting regions so I think it's not a huge deal?

**Comment by uklotzde** [comment]
I even think it is a good idea to start with relaxed settings:
* _ColumnLimit: 0_ to avoid eager reformatting
* _ReflowComments: false_ to disable reformatting of comments

Even with both options disabled the code should get more consistent and aligned with the coding style guidelines.

**Comment by uklotzde** [comment]
I've reverted all auto-formatting of files. This PR now only contains the new configuration file.

**Comment by ywwg** [comment]
thanks for sticking with this! we'll get through it :)

**Comment by uklotzde** [comment]
https://www.thoughtworks.com/radar/techniques/opinionated-and-automated-code-formatting

**Comment by uklotzde** [comment]
The current ThoughtWorks Technology Radar confirms my concerns and proposal and I strongly agree with their analysis and conclusions.

I get very good results when reformatting some files that I have previously edited with a customized Eclipse formatter. The great majority changes fix violations of our current coding style, e.g. indentation.

Using a common, imperfect code formatter would greatly help to keep the code base consistent and ease collaboration. I'm tired of arguing and refuse to discuss any coding style issues in manually formatted code in the future.

**Comment by daschuer** [comment]
Hey, it is no need to argue anymore. I think we all agree that this branch is a big step forward. 
The remaining question is about the steps to take. Even here we have an agreement for my feeling:

1.) Adjust our style guide, for a matching the clang-format rules.

2.) Clean PRs: With clang-format-diff or such touching changed lines only.

3.) Gain experiences before introducing the next steps above.

How is the current state regarding 1.) ?

**Comment by uklotzde** [comment]
I have updated and fixed the style guide, just some minor changes.

It doesn't make sense to describe all the formatting options in every detail that are already defined by the .clang-format configuration file. Maintaining consistency soon becomes tedious and is almost impossible to verify.

The style guide should only describe higher-level coding style or C++ design rules. The configuration file for formatting the code should just be referenced with a description on how to use or apply it. But this can't be done before the PR has been merged.

**Comment by daschuer** [comment]
Thank you for the work. I have two questions to the WIKI changes: 

> Use C++-style ​trailing line comments only. All trailing line comments start with a single space after the separator. Do not use C-style comments or Java-style comments.

This reads as if we cant do block of comments anymore. I think this was not your intention. Can we clarify this? 

> Avoid to comment out code unless a preceding textual comment explains the exact purpose of the following lines and why this code needs to be preserved as a comment. 

We have a lot of commented debug statements in the code I like to keep them without extra description. It is convenient to activate them in place.

**Comment by daschuer** [comment]
This is now ready to go. Thank you very much for keeping this ball up. 
LGTM

**Comment by Be-ing** [comment]
Please document on the wiki how to use this new tool. Maybe adding a new section to https://mixxx.org/wiki/doku.php/developer_guide would be a good place.

**Comment by uklotzde** [comment]
@daschuer I have updated the section about comments for clarification.

**Comment by uklotzde** [comment]
https://mixxx.org/wiki/doku.php/coding_guidelines#code_formatting

**Comment by daschuer** [comment]
I don't want to review mass formated code.
We can on each day mass format all code if the clang rules have settled.

Currently the wiki might encourage to do just this.

I would like remove all references to format whole files or directories. Instead we should explain how to use clang diff and set the IDE in a way that saved files are not Autoformated.

All style changing commits school be separated.

If you agree I can change the wiki accordingly.

**Comment by daschuer** [comment]
git-clang-format works great!
This is what should be used, not plain clang-format on whole files.  
I have edited the wiki page accordingly.

**Comment by ywwg** [comment]
this is so great! thank you @uklotzde

**Comment by uklotzde** [comment]
@daschuer Thank you for adding additional instructions to the wiki. I agree that encouraging developers to enable *Auto format on save* might cause a lot of noise. Such a recommendation should be deferred until some later stage when we might add even more automation like Git pre-commit hooks.

**Comment by ywwg** [comment]
One exercise that might be useful is finding source files that haven't been edited in a long time.  I bet we have some files that are quite stable, or only get edited when we do rename refactors.  Those files could be autoformatted with a low likelihood of messing up people's in-progress PRs.

If we can find some way to reformat code over time and minimize disruption that would be nice.  But I do think I agree that our style is pretty good and there isn't a lot of gross code that we *have* to fix

**Pull request merged**
